modules/post/windows/gather/checkvm.rb

```diff
@@ -9,349 +9,222 @@ class MetasploitModule < Msf::Post
 
-    if not vm
-      srvvals = registry_enumkeys('HKLM\SYSTEM\ControlSet001\Services')
-      if srvvals and srvvals.include?("vmicexchange")
-        vm = true
-      else
-        key_path = 'HKLM\HARDWARE\DESCRIPTION\System'
-        systemBiosVersion = registry_getvaldata(key_path,'SystemBiosVersion')
-        if systemBiosVersion.unpack("s<*").reduce('', :<<).include? "Hyper-V"
-          vm = true
-        end
-      end
-    end
+    srvvals = registry_enumkeys('HKLM\HARDWARE\ACPI\FADT')
+    return true if srvvals && srvvals.include?('VRTUAL')
 
-    if not vm
-      key_path = 'HKLM\HARDWARE\DEVICEMAP\Scsi\Scsi Port 0\Scsi Bus 0\Target Id 0\Logical Unit Id 0'
-      if registry_getvaldata(key_path,'Identifier') =~ /Msft    Virtual Disk    1.0/i
-        vm = true
-      end
-    end
+    srvvals = registry_enumkeys('HKLM\HARDWARE\ACPI\RSDT')
+    return true if srvvals && srvvals.include?('VRTUAL')
 
-    if vm
-      report_note(
-        :host   => session,
-        :type   => 'host.hypervisor',
-        :data   => { :hypervisor => "MS Hyper-V" },
-        :update => :unique_data
-        )
-      if physicalHost
-        print_good("This is a Hyper-V Virtual Machine running on physical host #{physicalHost}")
-      else
-        print_good("This is a Hyper-V Virtual Machine")
-      end
-      return "MS Hyper-V"
-    end
+    return true if get_services && get_services.include?('vmicexchange')
+
+    key_path = 'HKLM\HARDWARE\DESCRIPTION\System'
+    system_bios_version = registry_getvaldata(key_path, 'SystemBiosVersion')
+    return true if system_bios_version && system_bios_version.unpack('s<*').reduce('', :<<).include?('Hyper-V')
+
+    key_path = 'HKLM\HARDWARE\DEVICEMAP\Scsi\Scsi Port 0\Scsi Bus 0\Target Id 0\Logical Unit Id 0'
+    return true if registry_getvaldata(key_path, 'Identifier') =~ /Msft    Virtual Disk    1.0/i
+
+    false
   end
 
-  # Method for checking if it is a VMware VM
-  def vmwarechk(session)
-    vm = false
-    srvvals = registry_enumkeys('HKLM\SYSTEM\ControlSet001\Services')
-    if srvvals and  srvvals.include?("vmdebug")
-      vm = true
-    elsif srvvals and srvvals.include?("vmmouse")
-      vm = true
-    elsif srvvals and srvvals.include?("VMTools")
-      vm = true
-    elsif srvvals and srvvals.include?("VMMEMCTL")
-      vm = true
-    end
-    if not vm
-      if registry_getvaldata('HKLM\HARDWARE\DESCRIPTION\System\BIOS','SystemManufacturer') =~ /vmware/i
-        vm = true
-      end
-    end
-    if not vm
-      key_path = 'HKLM\HARDWARE\DEVICEMAP\Scsi\Scsi Port 0\Scsi Bus 0\Target Id 0\Logical Unit Id 0'
-      if registry_getvaldata(key_path,'Identifier') =~ /vmware/i
-        vm = true
-      end
+  def vmware?
+    if get_services
+      return true if get_services.include?('vmdebug')
+      return true if get_services.include?('vmmouse')
+      return true if get_services.include?('VMTools')
+      return true if get_services.include?('VMMEMCTL')
```

Comment:
This isn't super important, but I noticed that this pattern appears a few times in the module. Perhaps this could be replaced with something like `valid_service?`:

```ruby
def valid_service?(services)
  services.each { |service| return true if get_services.include?(service) }

  false
end
```

Reply:
done. i haven't tested the changes.

Reply:
Thanks! LGTM. I'll go ahead and test it out.